src/Microsoft.Health.Dicom.Blob/Features/Storage/BlobFileStore.cs

```diff
@@ -75,6 +78,85 @@ public class BlobFileStore : IFileStore
         }
     }
 
+    /// <inheritdoc />
+    public async Task<Uri> StoreFileInBlocksAsync(
+        long version,
+        Stream stream,
+        IDictionary<string, long> blockLengths,
+        CancellationToken cancellationToken)
+    {
+        EnsureArg.IsNotNull(stream, nameof(stream));
+        EnsureArg.IsNotNull(blockLengths, nameof(blockLengths));
+        EnsureArg.IsGte(blockLengths.Count, 0, nameof(blockLengths.Count));
+
+        BlockBlobClient blobClient = GetInstanceBlockBlobClient(version);
+
+        int maxBufferSize = (int)blockLengths.Max(x => x.Value);
+        byte[] buffer = ArrayPool<byte>.Shared.Rent(maxBufferSize);
+
+        try
+        {
+            foreach (string blockId in blockLengths.Keys)
+            {
+                long blockSize = blockLengths[blockId];
+
+                await stream.ReadAsync(buffer.AsMemory(0, (int)blockSize), cancellationToken);
+
+                using var blockStream = new MemoryStream(buffer);
+                await blobClient.StageBlockAsync(blockId, blockStream, cancellationToken: cancellationToken);
```

Comment:
can this be parallelized similar to Azure client sdk? Like a configurable 4/5 limit and upload all the stages in parallel and commit?

Another option is to use Azure client SDK with the blocksize option. For that all the blocks need to be same.

Reply:
This method is already called in parallel from a activity. 5 threads are parallely calling this method. Should we measure now and add the improvements later?

Reply:
Also the committed blocks are in the order similar to order we commit. If I make this parallel, we may need to preserve the first block name somewhere so we know how to retrieve it